src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,401 @@
+
+        let audience = self.config.audience.clone();
+        let subject_token_type = self.config.subject_token_type.clone();
+        let url = self.config.token_url.clone();
+        let mut scope = vec![];
+        if let Some(scopes) = self.config.scopes.clone() {
+            scopes.into_iter().for_each(|v| scope.push(v));
+        }
+        if scope.is_empty() {
+            scope.push(DEFAULT_SCOPE.to_string());
+        }
+        let req = ExchangeTokenRequest {
+            url,
+            audience: Some(audience),
+            subject_token,
+            subject_token_type,
+            scope,
+            authentication: ClientAuthentication {
+                client_id: self.config.client_id.clone(),
+                client_secret: self.config.client_secret.clone(),
+            },
+            ..ExchangeTokenRequest::default()
+        };
+
+        let token_res = STSHandler::exchange_token(req).await?;
+
+        let token = Token {
+            token: token_res.access_token,
+            token_type: token_res.token_type,
+            expires_at: Some(Instant::now() + Duration::from_secs(token_res.expires_in)),
+            metadata: None,
+        };
+        Ok(token)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ExternalAccountCredentials<T>
+where
+    T: CachedTokenProvider,
+{
+    token_provider: T,
+    quota_project_id: Option<String>,
+}
+
+/// A builder for external account [Credentials] instances.
+///
+/// # Example
+/// ```
+/// # use google_cloud_auth::credentials::external_account::{Builder};
+/// # tokio_test::block_on(async {
+/// let project_id = project_id();
+/// let workload_identity_pool_id = workload_identity_pool();
+/// let provider_id = workload_identity_provider();
+/// let provider_name = format!(
+///     "//iam.googleapis.com/projects/{project_id}/locations/global/workloadIdentityPools/{workload_identity_pool_id}/providers/{provider_id}"
+/// );
+/// let config = serde_json::json!({
+///     "type": "external_account",
+///     "audience": provider_name,
+///     "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+///     "token_url": "https://sts.googleapis.com/v1beta/token",
+///     "credential_source": {
+///         "url": format!("http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource={provider_name}"),
+///         "headers": {
+///           "Metadata": "True"
+///         },
+///         "format": {
+///           "type": "json",
+///           "subject_token_field_name": "access_token"
+///         }
+///     }
+/// });
+/// let credentials = Builder::new(config)
+///     .with_quota_project_id("quota_project")
+///     .build();
+/// });
+///
+/// fn project_id() -> String {
+///     "test-only".to_string()
+/// }
+/// fn workload_identity_pool() -> String {
+///     "test-only".to_string()
+/// }
+/// fn workload_identity_provider() -> String {
+///     "test-only".to_string()
+/// }
```

Comment:
You can hide these functions to make your example more readable in the rendered documentation:

```suggestion
/// # fn project_id() -> String {
/// #     "test-only".to_string()
/// # }
/// # fn workload_identity_pool() -> String {
/// #     "test-only".to_string()
/// # }
/// # fn workload_identity_provider() -> String {
/// #     "test-only".to_string()
/// # }
```

Alternatively, just hide their implementation.

Reply:
done